rfc/002-fault-tolerance/README.md

```diff
@@ -0,0 +1,487 @@
+consume the latest updates to synchronize the local representation of the state.
+The original key space that tracks the set of existing partitions would only
+serve to verify the validity of the transaction on request.
+
+To get a feel for the system, we briefly experimented with mapping our API to
+FDB primitives. We found the following:
+
+* Transactions are “all or nothing”, i.e., you can sequence multiple operations,
+  and if you commit and it goes through, they are all applied, otherwise rolled
+  back.
+
+* Multiple transactions that modify (read to or write from) the same key are not
+  forbidden, even if they overlap - to enable our use case, we’d have to verify
+  the existence of all the partitions we intend to delete, followed by doing all
+  the (SET, CLEAR) operations OR rollback / cancel the transaction.
+
+* While this strategy allows for serializing the transactions into a
+  conflict-free state, it does not help with each VAST node "seeing the full
+  picture", i.e., being aware of the transactions that are triggered from within
+  other VAST nodes.
+
+* There doesn’t seem to be a full CDC (change data capture) functionality, only
+  watching single keys - there could be a key for a "transaction_id" (an integer
+  increment for every transaction), and from there we could work around the
+  problem by requesting all recent transactions whenever this key is modified.
+
+##### etcd
+
+Etcd is a distributed key-value store that uses Raft for quorum-based consensus.
+From a CAP Theorem perspective, this brings us into the CP realm, because a
+master must be elected to perform writes. Stale reads may be supported, though,
+which we can tolerate.
+
+Here are some incomplete notes of taking a first look:
+
+- Multi write transactions are shaped like if-then-else statements, they always
+  have a comparison, a list of writes on success and a list of writes on failure
+
+- Watches are possible on single keys and prefixes in the key-space, which would
+  allow us to model our use-case naturally. Importantly, key updates are
+  streamed to the client, allowing a gap-free observation of changes in the
+  key-space.
+
+- Etcd comes with a gRPC API, making it easy to embed for us. We're already
+  depending on gRPC transitively via Apache Arrow (cf. Flight).
+
+- An official C++ client library
+- exists:https://github.com/etcd-cpp-apiv3/etcd-cpp-apiv3.
+##### Consul
+
+Consul is a service discovery and distributed configuration system that contains
+a distributed key-value store (consul KV). Apart from the bare key-value store
+it also provides service mesh, traffic management, and service to service
+encryption. For the use case at hand we identified the following relevant
+properties:
+
+- The only way to interact with a Consul service is through an HTTP API, meaning
+- there would be some overhead for each transaction. This also means that we
+- would have to use HTTP long polling to observe changes on a specific key.
```

Comment:
And that is bad because...?

Reply:
It is neither "good" nor "bad", but important to know nonetheless. The last point in this list explains the limitations that cone with a plain HTTP API.

Reply:
Comment that this limits us in an annoying way.

Reply:
I actually took out the last sentence and edited the last bullet instead.